src/lib/ams.c

```diff
@@ -0,0 +1,749 @@
+}
+
+int get_output_async_pin_props(struct async_message_service *ams, uint16_t module_id,
+			       uint16_t instance_id,
+			       uint32_t max_size,
+			       uint32_t *output_count,
+			       struct async_pin_props *pin_props)
+{
+	struct am_service_shared_context *shared_ctx = ams->am_service_context->shared;
+	struct ams_producer *producer_table;
+	uint32_t count = 0;
+	int32_t index;
+	int err = 0;
+
+	if (pin_props == NULL)
+		return -EINVAL;
+
+	ams_acquire(ams->am_service_context);
+	producer_table = shared_ctx->producer_table;
+
+	for (uint32_t i = 0; i < shared_ctx->producer_table_size; i++) {
+		if ((producer_table[i].producer_instance_id == instance_id) &&
+		    (producer_table[i].producer_module_id == module_id)) {
+			if (count == max_size) {
+				err = -EINVAL;
+				break;
+			}
+
+			index = am_service_find_uuid_index_by_msg_type_id(ams->am_service_context,
+									  producer_table[i].message_type_id);
+			if (index >= 0) {
+				memcpy_s(&pin_props[count].message_type_uuid, AM_MESSAGE_UUID_SIZE,
+					 &shared_ctx->uuid_table[index].message_uuid,
+					 AM_MESSAGE_UUID_SIZE);
+				count++;
+			} else {
+				tr_err(&ams_tr, "get_output_async_pin_props(): uuid index not found for message_type_id: %u",
+				       producer_table[i].message_type_id);
+			}
+		}
+	}
+
+	ams_release(ams->am_service_context);
+	if (err == 0)
+		*output_count = count;
+
+	return err;
+}
+
+int am_service_process_slot(struct async_message_service *ams, uint32_t slot)
+{
+	struct ams_message_payload *msg = &ams->am_service_context->shared->slots[slot].u.msg;
+	uint16_t module_id = ams->am_service_context->shared->slots[slot].module_id;
+	uint16_t instance_id = ams->am_service_context->shared->slots[slot].instance_id;
+
+	dcache_invalidate_region(&ams->am_service_context->shared->slots[slot], AMS_SLOT_SIZE(msg));
+
+	tr_info(&ams_tr, "am_service_process_slot slot %d msg %d from 0x%08x",
+		slot, msg->message_type_id,
+		msg->producer_module_id << 16 | msg->producer_instance_id);
+	return am_service_message_send_internal(ams, msg, module_id, instance_id, slot);
+}
+
+#if CONFIG_SMP
+
+void ams_task_add_slot_to_process(struct am_service_task *ams_task, uint32_t slot)
+{
+	int flags;
+
+	flags = arch_interrupt_global_disable();
+	ams_task->pending_slots |= BIT(slot);
+	arch_interrupt_global_enable(flags);
+}
+
+int process_incoming_message(uint32_t slot)
+{
+	int ret;
+	struct async_message_service *ams = sof_get()->ams;
+	struct am_service_task *task = &ams->ams_task;
+
+	ams_task_add_slot_to_process(task, slot);
+	ret = schedule_task(&task->ams_task, 0, 100);
+
+	return ret;
+}
+
+int process_slot(struct async_message_service *ams, uint32_t slot)
+{
+	return am_service_process_slot(ams, slot);
+}
+#endif /* CONFIG_SMP */
+
+/* ams task */
+
+enum task_state process_message(void *arg)
+{
+	struct am_service_task *ams_task = arg;
+	uint32_t slot;
+	int flags;
+
+	ams_task->is_in_do_work = true;
+	slot = (31 - XT_NSAU((ams_task->pending_slots)));
```

Comment:
double internal parentheses as well as the external ones are superfluous

Reply:
sure. I should also probably add xtensa agnostic version there. Should it be implemented in place there or do we have a separate way of doing that?